lean_source/06_Abstract_Algebra/source_01_Structures.lean

```diff
@@ -0,0 +1,503 @@
+``my_point1`` is a ``point``, you can start the definition by
+writing ``{! !}``. Clicking on the light bulb
+that appears nearby in VS Code will then
+give you the option of inserting a template definition
+with the field names listed for you.
+
+By default, the *constructor* for the ``point`` structure is named
+``point.mk``. You can specify a different name if you want.
+EXAMPLES: -/
+-- QUOTE:
+structure point' := build :: (x : ℝ) (y : ℝ) (z : ℝ)
+
+#check point'.build 2 (-1) 4
+-- QUOTE.
+
+/- TEXT:
+The next two examples show how to define functions
+on structures using pattern matching.
+Whereas the second example makes the ``point.mk``
+constructor explicit, the first example uses anonymous constructors
+for brevity.
+Lean can infer the relevant constructor from the indicated type of
+``add``.
+It is conventional to put definitions and theorems associated
+with a structure like ``point`` in a namespace with the same name.
+In the example below, the full name of ``add`` is ``point.add``,
+which allows us to use the anonymous projection notation.
+BOTH: -/
+-- QUOTE:
+namespace point
+
+def add : point → point → point
+| ⟨x₁, y₁, z₁⟩ ⟨x₂, y₂, z₂⟩ := ⟨x₁ + x₂, y₁ + y₂, z₁ + z₂⟩
+
+-- EXAMPLES:
+def add' : point → point → point
+| (point.mk x₁ y₁ z₁) (point.mk x₂ y₂ z₂) :=
+  { x := x₁ + x₂,
+    y := y₁ + y₂,
+    z := z₁ + z₂ }
+
+#check add my_point1 my_point2
+#check my_point1.add my_point2
+
+end point
+
+#check point.add my_point1 my_point2
+#check my_point1.add my_point2
+-- QUOTE.
+
+/- TEXT:
+Below we will continue to put definitions in the relevant
+namespace, but we will leave the namespacing commands out of the quoted
+snippets.
+
+Within a proof, structures can be decomposed with the ``cases`` or ``rcases``
+tactics. You can also use ``rintros`` or a pattern-matching
+lambda.
+
+EXAMPLES: -/
+namespace point
+
+-- QUOTE:
+theorem add_comm (a b : point) : add a b = add b a :=
+begin
+  cases a with xa ya za,
+  cases b with xb yb zb,
+  rw [add, add],
+  ext; dsimp,
+  repeat { apply add_comm }
+end
+
+example (a b : point) : add a b = add b a :=
+begin
+  rcases a with ⟨xa, ya, za⟩,
+  rcases b with ⟨xb, yb, zb⟩,
+  simp [add, add_comm]
+end
+
+example : ∀ a b : point, add a b = add b a :=
+begin
+  rintros ⟨xa, ya, za⟩ ⟨xb, yb, zb⟩,
+  simp [add, add_comm]
+end
+
+example : ∀ a b : point, add a b = add b a :=
+λ ⟨xa, ya, za⟩ ⟨xb, yb, zb⟩, by simp [add, add_comm]
+-- QUOTE.
+
+/- TEXT:
+In fact, even proving that the addition function does
+the right thing on the ``x`` coordinate
+requires decomposing the instances with ``cases``.
+EXAMPLES: -/
+-- QUOTE:
+theorem add_x (a b : point) : (a.add b).x = a.x + b.x :=
+by { cases a, cases b, refl }
+-- QUOTE.
+
+/- TEXT:
+Starting a proof with ``cases`` is a common pattern when dealing with functions
+that were defined by pattern matching.
+An alternative is to define functions using projections,
+in which case, the relevant equations often hold definitionally.
```

Comment:
Do we have anything convincing we could tell about pattern matching here? It sounds like using projections is both easier and leading to a nicer result. I'm afraid our readers will only remember that pattern matching is what functional programmers love for no good reason. 

By the way, should be include some comment about similarities and differences with object-oriented programming? All the examples from that page would clearly fit very well with classes in python say. And we'll have to face the name-clash between classes in Lean and classes in C++ or python sooner or later. Life would be so much simpler if classes in Lean could use another name...

Reply:
I also added this as a TODO. I worry about overloading the reader when there is already so much technical syntactic stuff in this chapter (especially in the next section, which I am about to PR). Let's revisit this when we can step back and see how the chapter reads.